### DISCUSSION TITLE
PID in Berry script

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21057] [Q&A]
Is there an option to use or implement a PID controller in a Berry script?
I want to control a fan with PWM based on a sensor value.

**Comment by wjohn007** [comment]
I don't have much experience with this repository [microserre](https://github.com/danjperron/microserre).
However, there is an implementation of PID using Berry.
Maybe that's helpful

**Reply by mabnz** [reply]
Where is the implementation of PID in Berry?

**Reply by s-hadinger** [reply]
It's not native in Tasmota. Maybe someone implemented it as Berry code

**Reply by sfromis** [reply]
According to the description in the linked `microserre` repository, [HeaterPWM.be](https://github.com/danjperron/microserre/blob/main/HeaterPWM.be) (along with [HeaterPWM_UI.be](https://github.com/danjperron/microserre/blob/main/HeaterPWM_ui.be)) has a PID implementation in Berry.

Additionally, https://github.com/glmoritz/tasmota_berry_pid has the about text starting with
> A Heat/Cool PID controller written in berry ...

https://github.com/darrylb123/tasmota/blob/main/ESP32-2432-S028/pid.be also seems to be in that direction....
